<html>
<head>
<title>
result help
</title>
</head>
<body>
<h3 align="center">help</h3>
<ul>
    <li>Please click <a href="FAQ.html#design">here</a> for the rule on selecting good siRNAs.
    <li>To reduce the non-specific target effects, you can <a href="FAQ.html#offtarget"">filter the siRNAs</a> by <a href="FAQ.html#identity">number of matches</a> and <a href="FAQ.html#position">alignment positions</a> of its BLAST result. After you make the selection and click on the "filter button", the program will first group the database hits in each siRNA BLAST results by GeneID (if you choose RefSeq as BLAST database), uniGene cluster (if you choose UniGene as BLAST database), or Ensembl gene (if you choose Ensembl transcripts as BLAST database), then filter out the siRNAs based on your criteria. Because GeneID, UniGene Cluster and Ensembl annotate genes in different systems, you may want to compare the results with different BLAST databases.
    <li>By default, the siRNAs are sorted by <a href="thermodynamics.html">thermodynamic values</a>. With the select box on the left side, you can resort the siRNAs by any one of the following items: query position, type (siRNA pattern), gc%, thermodynamic values.</li>
    <li>After select your siRNAs, by clicking on the "get siRNAs" button on the left, you can retrieve the sense and antisense siRNAs in tabular format.</li>
    <li>The html table can be downloaded by clicking on the "all_table" link on the left frame.
</ul> 
</body>
</html>
